scripts/vault-push-approles.nix

```diff
@@ -0,0 +1,251 @@
+# Generate and push approles to vault
+{ writeShellScriptBin, jq, vault, coreutils, lib }:
+# Inputs: a flake with `nixosConfigurations`
+
+# Usage:
+# apps.x86_64-linux.vault-push-approles = { type = "app"; program = "${pkgs.vault-push-approles self}/bin/vault-push-approles"; }
+
+{ nixosConfigurations ? { }, ... }: rec {
+  # Overrideable functions
+  # Usage examples:
+  # pkgs.vault-push-approles self { approleCapabilities.aquarius-albali-borgbackup = [ "read" "write" ]; }
+  /* pkgs.vault-push-approles self (final: prev: {
+       approleCapabilitiesFor = { approleName, namespace, ... }@params:
+         if namespace == "albali" then [ "read" "write" ] else prev.approleCapabilitiesFor params;
+     })
+  */
+  # pkgs.vault-push-approles { } { extraApproles = [ { ... } ] }
+
+  # `final` contains fixed-point functions after applying user-supplied overrides
+  overrideable = final: {
+
+    # Approles to upload in addition to the ones generated from
+    # vault-secrets' NixOS module definitions. Must contain all the
+    # options from top-level vault-secrets option and all the options
+    # from vault-secrets.secrets.<name> submodule, as well as
+    # "approleName" attribute
+    extraApproles = [ ];
+
+    # Render an attrset into a JSON file
+    renderJSON = name: content:
+      builtins.toFile "${name}.json" (builtins.toJSON content);
+
+    # Default approle parameters
+    approleParams = {
+      secret_id_ttl = "";
+      token_num_uses = 0;
+      token_ttl = "20m";
+      token_max_ttl = "30m";
+      secret_id_num_uses = 0;
+    };
+
+    # Generate an approle parameters attrset based on its name and other
+    # options from its secret definition
+    mkApprole = { approleName, ... }:
+      (final.approleParams // { token_policies = [ approleName ]; });
+
+    # Create a JSON (HCL) file with approle parameters in it from its secret definition
+    renderApprole = { approleName, ... }@params:
+      final.renderJSON "approle-${approleName}" (final.mkApprole params);
+
+    # An attrset mapping `approleName`s to capabilities required by those approles
+    approleCapabilities = { };
+
+    # Get capabilities for the given secret definition
+    approleCapabilitiesFor = { approleName, ... }:
+      final.approleCapabilities.${approleName} or [ "read" ];
+
+    # Generate an approle policy from its secret definition
+    mkPolicy = { approleName, name, vaultPathPrefix, namespace, ... }@params:
+      let
+        splitPrefix = builtins.filter builtins.isString
+          (builtins.split "/" vaultPathPrefix);
+
+        insertAt = lst: index: value:
+          (lib.lists.take index lst) ++ [ value ] ++ (lib.lists.drop index lst);
+
+        makePrefix = value:
+          builtins.concatStringsSep "/" (insertAt splitPrefix 1 value);
+
+        metadataPrefix = makePrefix "metadata";
+        dataPrefix = makePrefix "+";
```

Comment:
I think if we just need to give read access, then the prefix can be just `data`  instead of matching everything with `+`

Reply:
We need to also give `write` in some cases, and I think matching everything for `read` should be fine?

Reply:
What do we need `write` for? We have it in policies we give to real people, but for approles used by vault-secrets we only need read access, don't we?

> I think matching everything for read should be fine?
Yeah, it probably is, it's just a bit confusing that we use a wildcard prefix when we only need this rule for `data`

Reply:
> What do we need write for? We have it in policies we give to real people, but for approles used by vault-secrets we only need read access, don't we?

AFAIU the plan is to rewrite everything to this script (using `extraApproles` for stuff not in `vault-secrets.secrets`), so we need to consider real people policies as well.

Reply:
I see, but for people we use OIDC instead of approles, so we would have to adapt this script somehow first

Reply:
> the plan is to rewrite everything to this script (using extraApproles for stuff not in vault-secrets.secrets), so we need to consider real people policies as well.

What? No. Let's keep this generator focused on providing AppRoles for servers.

Reply:
Hm, how would we upload people policies then?

Actually, a better question is what repo to put those policies in...

Reply:
Default choice would be serokell.nix, as the place where we put all of our "things" that are not for any specific constellation.

I can't think of any better place, unless you want to make a separate repo for it.

Reply:
I think serokell.nix is now bloated far beyond what it was originally intended for (not saying that's a bad thing). Maybe we could make serokell/team and put ssh&gpg pubkeys, vault approles, etc there, and use serokell/common-infra for stuff like what's currently in `common.nix` in `serokell.nix`. Then we could use `serokell.nix` in app repos without downloading tons of unused crap.

Reply:
On one hand, "tons of unused crap" is, what, ~3Kb? On the other hand, organizing things is never a bad thing. I've been meaning to do a bit of a cleanup in serokell.nix for a while now. It's not necessarily too big, but it's not neatly organized.